global/global.go

```diff
@@ -110,3 +116,20 @@ func (t *TLS) Config() (*tls.Config, error) {
 		Certificates: []tls.Certificate{certificate},
 	}, nil
 }
+
+// DoRetry is a help function to retry the function if it returns error
+func DoRetry(kind, name, tag string, r Retry, fn func() error) error {
+	for i := 0; i < r.Times; i++ {
+		err := fn()
+		if err == nil {
+			return nil
+		}
+		log.Warnf("[%s / %s / %s] Retried to send %d/%d - %v", kind, name, tag, i+1, r.Times, err)
+
+		// last time no need to sleep
+		if i < r.Times-1 {
+			time.Sleep(r.Interval)
+		}
+	}
+	return fmt.Errorf("[%s / %s / %s] failed after %d retries", kind, name, tag, r.Times)
```

Comment:
It would be better to print the detailed information of the last error

Reply:
done!